File: src/main/scala/com/thoughtworks/pli/remotepair/idea/dialogs/JoinProjectDialog.scala
package com.thoughtworks.pli.remotepair.idea.dialogs

import com.thoughtworks.pli.intellij.remotepair.protocol._
import com.thoughtworks.pli.remotepair.core._
import com.thoughtworks.pli.remotepair.core.client.{PublishEvent, GetServerWatchingFiles, GetExistingProjects}
import com.thoughtworks.pli.remotepair.core.models.MyPlatform
import com.thoughtworks.pli.remotepair.idea.idea.{ShowServerError, GetProjectWindow}
import com.thoughtworks.pli.remotepair.idea.listeners.PairEventListeners
import com.thoughtworks.pli.remotepair.idea.settings.ClientNameInGlobalStorage

import scala.collection.JavaConversions._

object JoinProjectDialog {
  type Factory = () => JoinProjectDialog
}

class JoinProjectDialog(val myPlatform: MyPlatform, watchFilesDialogFactory: WatchFilesDialog.Factory, val pairEventListeners: PairEventListeners, logger: PluginLogger, publishEvent: PublishEvent, showServerError: ShowServerError, getExistingProjects: GetExistingProjects, clientNameInGlobalStorage: ClientNameInGlobalStorage, val getProjectWindow: GetProjectWindow, getServerWatchingFiles: GetServerWatchingFiles) extends _JoinProjectDialog with JDialogSupport {

  onWindowOpened(initDialog())
  monitorReadEvent {
    case JoinedToProjectEvent(projectName, clientName) => chooseIgnoreFiles()
    case ProjectOperationFailed(msg) => showErrorMessage(msg)
  }
  onClick(okButton)(publishProjectEvent())

  private def initDialog(): Unit = {
    getExistingProjects().foreach(generateRadio)
    init()
    clientNameTextField.setText(clientNameInGlobalStorage.load())
  }

  private def chooseIgnoreFiles(): Unit = {
    this.dispose()
    if (getServerWatchingFiles().isEmpty) {
      watchFilesDialogFactory(None).showOnCenter()
    }
  }

  private def publishProjectEvent() = {
    clientNameInGlobalStorage.save(clientNameTextField.getText)
    errorMessageLabel.setVisible(false)
    try {
      getSelectedOrCreatedProjectName match {
        case Some(ExistingProjectName(p)) => publishEvent(new JoinProjectRequest(p, clientNameTextField.getText))
        case Some(NewProjectName(p)) => publishEvent(new CreateProjectRequest(p, clientNameTextField.getText))
        case _ => showErrorMessage("No valid project name")
      }
    } catch {
      case e: Throwable =>
        errorMessageLabel.setVisible(true)
        errorMessageLabel.setText(e.toString)
    }
  }

  private def getSelectedOrCreatedProjectName: Option[ProjectName] = {
    projectRadios.find(_.isSelected).map(_.getText) match {
      case Some(name) => Some(ExistingProjectName(name))
      case _ => Some(newProjectNameTextField.getText.trim).filter(_.nonEmpty).map(NewProjectName.apply)
    }
  }
}

sealed trait ProjectName
case class ExistingProjectName(name: String) extends ProjectName
case class NewProjectName(name: String) extends ProjectName

File: src/main/scala/com/thoughtworks/pli/remotepair/idea/dialogs/_JoinProjectDialog.java
package com.thoughtworks.pli.remotepair.idea.dialogs;

import com.thoughtworks.pli.remotepair.idea.core.ProjectWithMemberNames;

import javax.swing.*;
import javax.swing.event.ChangeEvent;
import javax.swing.event.ChangeListener;
import java.awt.*;
import java.util.ArrayList;
import java.util.List;

public class _JoinProjectDialog extends JDialog {

    private JPanel contentPanel;
    protected JRadioButton newProjectRadio;
    protected JTextField newProjectNameTextField;
    protected JPanel existingProjectPanel;
    protected JTextField clientNameTextField;
    protected JLabel errorMessageLabel;
    protected JButton okButton;
    private ButtonGroup buttonGroup = new ButtonGroup();
    protected List<JRadioButton> projectRadios = new ArrayList<JRadioButton>();

    public _JoinProjectDialog() {
        setContentPane(contentPanel);
    }

    protected void init() {
        errorMessageLabel.setVisible(false);
        this.existingProjectPanel.setLayout(new BoxLayout(existingProjectPanel, BoxLayout.Y_AXIS));
        buttonGroup.add(newProjectRadio);
        this.newProjectRadio.addChangeListener(new ChangeListener() {
            @Override
            public void stateChanged(ChangeEvent changeEvent) {
                boolean enable = newProjectRadio.isSelected();
                newProjectNameTextField.setEnabled(enable);
            }
        });

        if (projectRadios.isEmpty()) {
            newProjectRadio.setSelected(true);
            newProjectNameTextField.requestFocus();
        }
    }

    protected void generateRadio(ProjectWithMemberNames project) {
        JRadioButton radio = new JRadioButton(project.projectName());
        JLabel label = new JLabel(project.memberNames().mkString(" : ", ",", ""));
        existingProjectPanel.add(newPanel(radio, label));
        projectRadios.add(radio);
        buttonGroup.add(radio);
    }

    private JPanel newPanel(JRadioButton radio, JLabel label) {
        JPanel panel = new JPanel();
        panel.setLayout(new BoxLayout(panel, BoxLayout.X_AXIS));
        panel.setAlignmentX(Component.LEFT_ALIGNMENT);
        panel.add(radio);
        panel.add(label);
        return panel;
    }

    public void showErrorMessage(String message) {
        errorMessageLabel.setText("Error: " + message);
        errorMessageLabel.setVisible(true);
    }

    public void hidePreErrorMessage() {
        errorMessageLabel.setVisible(false);
    }

}
